File: src/main/kotlin/davidoh/profiling/SampleApp.kt
import davidoh.profiling.SimpleProfiler
import java.security.MessageDigest
import kotlin.math.sqrt

fun main(args: Array<String>) {
    val profiler = SimpleProfiler(reportSec = 10, enclosingSectionName = "total", resetAfterSampleCount = 500000)

    println("app started")

    profiler.startSection("init")
    val md5 = MessageDigest.getInstance("MD5")
    val sha1 = MessageDigest.getInstance("SHA-1")
    val sha256 = MessageDigest.getInstance("SHA-256")
    val r = java.util.Random()
    val bytes = ByteArray(10000)
    r.nextBytes(bytes)
    profiler.endSection()

    while (true) {
        profiler.startSection("total")

        profiler.startSection("hashing")

        profiler.startSection("md5")
        md5.digest(bytes)
        profiler.endSection()

        profiler.startSection("sha256")
        sha256.digest(bytes)
        profiler.endSection()

        profiler.startSection("sha1")
        sha1.digest(bytes)
        profiler.endSection()

        profiler.endSection("hashing")

        profiler.startSection("random")
        val randNum = r.nextDouble() * 1000000 + 1
        profiler.endSection()

        profiler.startSection("sqrt")
        sqrt(randNum)
        profiler.endSection()

//        profiler.startSection("probablePrime")
//        BigInteger.probablePrime(32, r)
//        profiler.endSection()

        profiler.endSection("total")
        if (profiler.periodicReport())
            println("-----")
    }
}

File: python/sample_app.py
import hashlib
import math
import os
import random

from simple_profiler import SimpleProfiler

print("app started")

profiler = SimpleProfiler(enclosing_section_name="total",
                          reset_after_sample_count=500000)

profiler.start_section("init")
payload = os.urandom(10000)
r = random.random()
profiler.end_section()

while True:
    profiler.start_section("total")

    profiler.start_section("hashing")

    profiler.start_section("md5")
    hashlib.md5(payload)
    profiler.end_section()

    profiler.start_section("sha1")
    hashlib.sha1(payload)
    profiler.end_section()

    profiler.start_section("sha256")
    hashlib.sha256(payload)
    profiler.end_section()

    profiler.end_section("hashing")

    profiler.start_section("random")
    rand_number = random.random() * 1000000 + 1
    profiler.end_section()

    profiler.start_section("sqrt")
    math.sqrt(rand_number)
    profiler.end_section()

    profiler.end_section("total")

    if profiler.report(10):
        print("-------")
